### DISCUSSION TITLE
Embedding non-embedded fonts

### DISCUSSION THREAD

**Discussion by sfadschm** [pymupdf/PyMuPDF-Utilities#135] [Q&A]
Hi,
I currently have a pdf file which has some embedded and some non-embedded fonts.
What I want to achieve is switching the non-embedded fonts to embedded without changing anything else in the file.

So far, I could not find any means to do that.
I found

-  https://github.com/pymupdf/PyMuPDF/issues/73, which might hint, that such action is not possible at all at current time.
- https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/font-replacement, which I believe might do the trick if I replace all non-embedded fonts with themselves, but has a high risk of breaking some layout due to rewriting the whole text of replaced fonts.

Is there any low-threshold method to just tell pdf to embed the font?

Side-note: E.g., in "Foxit PDF Editor", the only way to do so is to "print" the whole document with the internal printer again, checking the "embed all fonts" checkbox.

Thanks in advance!

**Comment by JorjMcKie** [comment]
No, and I believe this cannot ever work without actually rewriting the text.
Non-embedded fonts are "simple" fonts, where Unicodes are <= 255 and glyph number = Unicode number.
The embeddable Helvetica version works entirely different here compared to its Base-14 pendant.

**Reply by sfadschm** [reply]
Alright, thanks for the elaboration!

**Reply by Sahil-CE** [reply]
Hi, did you manage to find a fix for this?
I’m facing a similar issue where the PDF contains non-embedded fonts. If I already have a TTF file (which could be the same font, a replacement, or any other TTF), what’s the proper way to replace the non-embedded font reference in the PDF and embed this TTF in its place?

Thanks in advance!!

**Reply by JorjMcKie** [reply]
@Sahil-CE - Just to underpin my previous statement:
There is no way to "replace" a font and there never will be.
The only circumvention is to rewrite the PDF with a utility like shown [here](https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/font-replacement).

**Reply by sfadschm** [reply]
Not possible as explained. Workaround is to limit your character set/ use different fonts, but really most likely not worth the effort

**Reply by Sahil-CE** [reply]
@JorjMcKie thanks a lot for your reply and direction, I’ll definitely check out that solution.

I just had a follow-up question out of curiosity: when Adobe embeds fonts in an existing PDF, do they essentially rewrite the PDF under the hood, or is their approach different? I’m trying to understand how Adobe handles this process (embeds fonts in preprint) and whether there’s any known methodology or documentation around it. Do you happen to have any insight on that?

Thanks again for your time and help!

**Reply by Sahil-CE** [reply]
@sfadschm Thanks for the suggestion!!
